### PR TITLE
Chi-Square AB test power analysis

### PR DESCRIPTION
**Pull request by hideaki**
# Description
Chi-Square AB test power analysis.

# Checklist
Make sure you have performed the following items before submitting this pull request.
If not, please describe the reason.  

- [x] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/test_wrapper.R [hunk 1]
```diff
@@ -43,6 +43,29 @@ generate_chisq_density_data <- function(stat, df, sig_level = 0.05) {
   ret
 }
 
+generate_chisq_density_data_for_power <- function(df, w, N, crit) {
+  # Plot up to 95 percentile.
+  ncp <- N*w^2
+  l <- qchisq(0.95, df=df, ncp=ncp)
```
Comment (kei51e) [inline R/test_wrapper.R:49]: Add a comment for "0.95".

File: R/test_wrapper.R [hunk 2]
```diff
@@ -43,6 +43,29 @@ generate_chisq_density_data <- function(stat, df, sig_level = 0.05) {
...
+  x <- seq(from=0, to=l, by=l/1000 )
+
+  ret <- tibble::tibble(x=x,
+                        y=dchisq(x, df=df),
+                        type="Null"
+  )
+  ret0 <- tibble::tibble(x=crit, y=dchisq(crit, df=df), type="Null", statistic=TRUE)
```
Comment (kei51e) [inline R/test_wrapper.R:56]: Add comments.

File: R/test_wrapper.R [hunk 3]
```diff
@@ -2134,3 +2157,63 @@ tidy.shapiro_exploratory <- function(x, type = "model", signif_level=0.05) {
     ret
   }
 }
+
+#' @export
```
Comment (kei51e) [inline R/test_wrapper.R:2161]: Add a comment for 'dummy'.

File: R/test_wrapper.R [hunk 4]
```diff
@@ -2134,3 +2157,63 @@ tidy.shapiro_exploratory <- function(x, type = "model", signif_level=0.05) {
...
+exp_chisq_power <- function(dummy, rows=2, cols=2, w=0.3, sig.level=0.05, beta=0.2, n_start=10, n_end=100, n_step=10) {
+  power <- 1.0 - beta
+  n = seq(n_start, n_end, by=n_step)
+  # t-test
```
Comment (kei51e) [inline R/test_wrapper.R:2165]: Remove

File: R/test_wrapper.R [hunk 5]
```diff
@@ -2134,3 +2157,63 @@ tidy.shapiro_exploratory <- function(x, type = "model", signif_level=0.05) {
...
+                  power=power,
+                  required_n=required_n,
+                  density=density)
+    class(model) <- c("chisq_power_exploratory")
+    model
+  }
+  do_on_each_group(dummy, chisq_power_each, name = "model", with_unnest = FALSE)
+}
+
+#' @export
```
Comment (kei51e) [inline R/test_wrapper.R:2195]: Add a comment. Describe the difference from `exp_chisq_power`.

File: R/util.R [hunk 6]
```diff
@@ -2199,6 +2199,22 @@ calculate_cohens_w <- function(chi_sq, N) {
   sqrt(chi_sq/N)
 }
 
+# Calculate Cohen's w from the following input for 2x2 AB test case.
+# - Ratio of sample size between A and B
+# - Expected overall conversion rate
+# - % difference to detect. i.e. conversion(A) - conversion(B)
```
Comment (kei51e) [inline R/util.R:2205]: Update the comment. It is actually not "%" but "ratio".

File: R/util.R [hunk 7]
```diff
@@ -2199,6 +2199,22 @@ calculate_cohens_w <- function(chi_sq, N) {
...
+calculate_cohens_w_for_ab_test <- function(a_ratio, conversion_rate, diff) {
+  expected <- matrix(c(a_ratio, 1-a_ratio)) %*% matrix(c(conversion_rate, 1-conversion_rate), nrow = 1)
+  # Devide the diff into a_up and b_down, without changing the overall mean conversion rate.
```
Comment (kei51e) [inline R/util.R:2208]:
```suggestion
  # Divide the diff into a_up and b_down, without changing the overall mean conversion rate.
```

File: tests/testthat/test_chisq_power.R [hunk 8]
```diff
@@ -0,0 +1,27 @@
+context("test chisq power analysis")
+test_that("chisq AB test effect size calculation", {
+  res <- calculate_cohens_w_for_ab_test(0.5, 0.3, 0.02)
+  expect_equal(res, 0.02182179, tolerance=0.0001)
+})
+
+test_that("chisq AB test power analysis", {
+  model_df <- airquality %>% exp_chisq_power_for_ab_test(a_ratio=0.5, conversion_rate=0.1, diff=0.01, sig.level = 0.05, beta = 0.2)
+  expect_true(!is.null(model_df$model))
+  res <- model_df %>% tidy_rowwise(model, type="summary")
```
Comment (kei51e) [inline tests/testthat/test_chisq_power.R:10]: Check the output numbers.

File: R/test_wrapper.R [hunk 9]
```diff
@@ -43,6 +43,29 @@ generate_chisq_density_data <- function(stat, df, sig_level = 0.05) {
   ret
 }
 
+generate_chisq_density_data_for_power <- function(df, w, N, crit) {
```
[the remark this answers is not in the document]
Comment (hideaki) [reply R/test_wrapper.R:46]: Done.

File: R/test_wrapper.R (hunk 1 above)
Comment (hideaki) [reply R/test_wrapper.R:49]: Done.

File: R/test_wrapper.R (hunk 2 above)
Comment (hideaki) [reply R/test_wrapper.R:56]: Done.

File: R/test_wrapper.R (hunk 3 above)
Comment (hideaki) [reply R/test_wrapper.R:2161]: Done.

File: R/test_wrapper.R (hunk 5 above)
Comment (hideaki) [reply R/test_wrapper.R:2195]: Done.

File: R/util.R (hunk 6 above)
Comment (hideaki) [reply R/util.R:2205]: Done.

File: R/test_wrapper.R (hunk 4 above)
Comment (hideaki) [reply R/test_wrapper.R:2165]: Done.

File: tests/testthat/test_chisq_power.R (hunk 8 above)
Comment (hideaki) [reply tests/testthat/test_chisq_power.R:10]: Done.

**Review by kei51e — APPROVED** [review]

**Pull request merged**